service/src/main/kotlin/fi/espoo/evaka/shared/data/RangeBasedMap.kt

```diff
@@ -0,0 +1,152 @@
+    fun removeAll(ranges: Sequence<Range>): This = ranges.fold(this.entries, ::remove).toThis()
+
+    fun contains(range: Range): Boolean =
+        this.entries.fold(emptyList<Range>()) { acc, entry ->
+            entry.first.intersection(range)?.let { overlap -> RangeBasedSet.add(acc, overlap) }
+                ?: acc
+        } == listOf(range)
+
+    protected abstract fun List<Pair<Range, T>>.toThis(): This
+    protected abstract fun range(start: Point, end: Point): Range
+
+    companion object {
+        fun <T, Point : Comparable<Point>, Range : BoundedRange<Point, Range>> update(
+            entries: List<Pair<Range, T>>,
+            range: Range,
+            newValue: T,
+            resolve: (range: Range, old: T, new: T) -> T
+        ): List<Pair<Range, T>> {
+            val result = mutableListOf<Pair<Range, T>>()
+
+            fun Pair<Range, T>.tryMerge(other: Pair<Range, T>) =
+                if (this.first.adjacentTo(other.first) && this.second == other.second)
+                    this.first.merge(other.first) to this.second
+                else null
+
+            fun append(entry: Pair<Range, T>) =
+                when (val last = result.lastOrNull()) {
+                    null -> result.add(entry)
+                    else -> {
+                        when (val merged = last.tryMerge(entry)) {
+                            null -> result.add(entry)
+                            else -> {
+                                result.removeLast()
+                                result.add(merged)
+                            }
+                        }
+                    }
+                }
+
+            var newRange: Range? = range
+            for ((oldRange, oldValue) in entries) {
+                if (newRange == null || oldRange.strictlyLeftTo(newRange)) {
+                    append(oldRange to oldValue)
+                    continue
+                }
+                newRange =
+                    when (val overlap = oldRange.intersection(newRange)) {
+                        null -> {
+                            // If the old range is not on the left side and there's no overlap, it
+                            // must be on the right side. Since the old ranges are sorted, further
+                            // entries are guaranteed to not overlap.
+                            append(newRange to newValue)
+                            append(oldRange to oldValue)
+                            null
+                        }
+                        else -> {
+                            val subNew = newRange.subtract(overlap)
+                            val subOld = oldRange.subtract(overlap)
+
+                            // only one left range can be non-null, because otherwise there would
+                            // be more overlap
+                            subNew.left?.let { append(it to newValue) }
+                            subOld.left?.let { append(it to oldValue) }
```

Comment:
I might consider adding some invariant/sanity checks whenever they are cheap to fail visibly 
`if (subNew.left != null && subOld.left != null) throw Error("Bug: unexpected error")`

Reply:
We could even add `fun assertInvariants()` to RangeBasedSet and RangeBasedMap and call it on constructor to make sure that ranges don't overlap etc